Filter Menu Button

seangates opened the pull request:
## Description
- new `filter-menu-button` module
- fixes to `filter-button` and `filter-menu` as needed to support the new module

## References
Closes #701 

## Screenshots

### DS4

![image](https://user-images.githubusercontent.com/105656/62984949-640fc080-bdf2-11e9-921d-278db967b1bc.png)

![image](https://user-images.githubusercontent.com/105656/62984967-7984ea80-bdf2-11e9-849f-f5aefac429ea.png)

![image](https://user-images.githubusercontent.com/105656/62984977-873a7000-bdf2-11e9-916d-e6ed5cdfa7cb.png)

### DS6

![image](https://user-images.githubusercontent.com/105656/62985025-a89b5c00-bdf2-11e9-86a9-be61823b0ec2.png)

![image](https://user-images.githubusercontent.com/105656/62985038-ba7cff00-bdf2-11e9-8f9d-0aa73e07028f.png)

![image](https://user-images.githubusercontent.com/105656/62985059-d385b000-bdf2-11e9-8d23-adf76ede4489.png)

msendlakowski commented on docs/_includes/common/filter-menu.html:
```diff
@@ -1,8 +1,8 @@
-    <h3 id="filter-menu">Filter Menu – ARIA Version – Default</h3>
-    <p>The ARIA version of the filter menu is a <em>statelful</em> menu. It's primary purpose is used in conjunction with <a href="#filter-menu-button">filter-menu-button</a>,
+    <h3 id="filter-menu">Filter Menu – Default</h3>
+    <p>The default version of the filter menu is a <em>statelful</em> menu. It's primary purpose is used in conjunction with <a href="#filter-menu-button">filter-menu-button</a>,
```
`statelful`

seangates replied:
Wish this were a suggestion and then I could just take the suggestion. 😉 

Will fix.

msendlakowski commented on docs/src/js/base/main.js:
```diff
@@ -202,6 +202,28 @@ document.querySelectorAll('.menu-button').forEach(function(widgetEl) {
 });
 
+document.querySelectorAll('.filter-menu-button').forEach(function(widgetEl) {
+    console.log('**** widgetEl ***********************************************');
```
Debugging statements left in?

seangates replied:
D'oh! Will fix.

msendlakowski commented on src/less/filter-menu-button/base/filter-menu-button.less:
```diff
@@ -0,0 +1,247 @@
+    .background-icon-image-mixin('checkbox-checked', 18px, 18px);
+}
+
+.filter-menu-button__item[role="menuitemcheckbox"][aria-checked="true"] > span.filter-menu-button__status--inline > svg.icon.icon--checkbox-unchecked {
```
(minor) This and the next selector have a lot in common except for the last part.  They could be combined.

seangates replied:
@msendlakowski You're proposing something like this?:

```
.filter-menu-button__item[role="menuitemcheckbox"][aria-checked="true"]>span.filter-menu-button__status--inline>svg.icon.icon--checkbox {
    &-unchecked {
        display: none;
    }

    &-checked {
        display: block;
    }
}
```

msendlakowski replied:
No, cause I'd be reprimanded for that.  I'm thinking:
```
.filter-menu-button__item[role="menuitemcheckbox"][aria-checked="true"]>span.filter-menu-button__status--inline {
    > svg.icon.icon--checkbox-unchecked {
        display: none;
    }

    > svg.icon.icon--checkbox-checked {
        display: block;
    }
}
```

seangates replied:
Ha ha. Yeah, that makes sense.

ianmcburnie commented:
Now that we have the new test page setup, I think you could move a couple of these examples under the test. Like "many options" and "truncated". Also please try and add a short one-liner before each example to describe the use case a bit more and/or what modifiers/attributes must be applied.

ianmcburnie commented:
Also, I don't see any example code in those screenshots...

ianmcburnie commented:
@seangates So is this just the ARIA version currently? It's not clear from the description. If so, will there be a version that opens a filter-menu-form too?

agliga commented on src/less/filter-menu-button/base/filter-menu-button.less:
```diff
@@ -0,0 +1,247 @@
+    position: relative;
+}
+
+span.filter-menu-button__items {
```
Could we change the __items to something else, like menu or something. It was really confusing trying to discern between __item and __items. I know we did the same with listbox and such and it's much cleaner.
I'm fine with doing it as a refactor if it's too complex.

seangates replied:
@ianmcburnie This followed your `menu` pattern. Should we consider changing it in both places.?

ianmcburnie replied:
Let's leave as is for now. I'll have a think.

agliga commented on src/less/filter-menu-button/base/filter-menu-button.less:
```diff
@@ -0,0 +1,247 @@
+
+button.filter-menu-button__button[disabled],
+button.filter-menu-button__button[aria-disabled="true"],
+button.filter-menu-button__button[aria-pressed="true"][disabled],
```
Do we need all the extra selectors? I would think [disabled] and [aria-disabled=true] are good enough. Why need the aria-pressed too?

seangates replied:
Those are the different variations of states of the button. If I leave one out then we have a situation where it doesn't properly style for that combination of properties.

agliga replied:
Won't button.filter-menu-button__button[disabled] selector be applied when aria-pressed=true? Or am I missing something. If so, then we don't need [aria-pressed=true][disabled].

seangates replied:
There are conflicting styles with when `[aria-pressed="true"]` and just `[disabled]` given the order of the CSS in the file (avoiding the `no-descending-specificy` rule we have set up).

seangates replied:
But ... I will look into simplifying this.

agliga replied:
Got it. Just use `!important`!....

Just kidding, that's what I assumed but just wanted to double check!

seangates commented:
@ianmcburnie The example code is there, I just hid it for the screen shots.

seangates commented:
@ianmcburnie @agliga @msendlakowski I added a "form" version of the menu. I also updated the documentation a bit, and added the test pages. Please review again.

msendlakowski commented:
The form version isn't keyboard accessible.  

Also, I thought forms without submit buttons are an a11y no-no.  Should our demo include the "Apply" footer button?

seangates commented:
@msendlakowski It should be accessible, as it uses tabs instead of arrow keys.

Also, it should have a submit button: `type="submit"`. Some copy/pasta there. Fixing.

seangates commented:
@msendlakowski Looks like there might be an issue in `makeup-roving-tabindex` preventing spacebar from checking the checkbox in the form version. I'll have to dig into that.

/cc @ianmcburnie

ianmcburnie commented:
A roving tab index isn't strictly necessary with the form version. Kind of similar to the fake-menu.

ianmcburnie commented on docs/_includes/tests/filter-menu-button.html:
```diff
@@ -0,0 +1,650 @@
+                <div class="filter-menu-button__menu">
+                    <div class="filter-menu-button__items" role="menu">
+                        <div class="filter-menu-button__item" role="menuitemcheckbox" aria-checked="false">
+                            <span class="filter-menu-button__status--inline">
```
@seangates Sorry, but I just spotted this `filter-menu-button__status--inline`. This is actually invalid BEM. You can't have a modifier class without a base class.

seangates replied:
A modifier is invalid BEM?

ianmcburnie replied:
Yup. A modifier is supposed to modify a base class. In this case, `filter-menu-button__status`

msendlakowski replied:
I was wondering about that when I saw it too, since I'd never seen it before.  But according to the BEM docs modifiers are `Flags on blocks or elements.`
So `filter-menu-button__status` would be the element and `--inline` would be the modifier, right?

msendlakowski replied:
But I'm assuming there is a `filter-menu-button__status` element without the modifier.  So I guess basically there is no need for the modifier if there is no "unmodified" version of the element.

ianmcburnie replied:
If you only have one class, what other class styles are you modifying in the cascade?

msendlakowski replied:
Right, so it would have to be `filter-menu-button__status filter-menu-button__status--inline` but there's no need for the modifier if it isn't actually modifying anything from `filter-menu-button__status`.  Got it.

ianmcburnie commented on docs/_includes/tests/filter-menu-button.html:
```diff
@@ -0,0 +1,650 @@
+                <button type="button" class="filter-menu-button__button">
+                    <span class="filter-menu-button__button-cell">
+                        <span class="filter-menu-button__button-text">Filter Menu Button</span>
+                        <svg class="icon icon--chevron-down" focusable="false" height="12" width="12">
```
@seangates @msendlakowski We must also support the background svg use case for the button, e.g. `<span class="filter-menu-button__icon"><span>' instead of <svg> tag.

seangates replied:
We do. I just did not provide an example of it.

ianmcburnie replied:
I tried it and it did not work.

ianmcburnie commented on docs/_includes/common/filter-menu-button.html:
```diff
@@ -0,0 +1,654 @@
+                                <span class="filter-menu-button__text">Item 1 with a long string that will wrap to another line</span>
+                            </div>
+                            <div class="filter-menu-button__item" role="menuitemcheckbox" aria-checked="false">
+                                <span class="filter-menu-button__status--inline">
```
@seangates @msendlakowski so this got merged with the invalid BEM still present? :-( A base class needs to be present, or this modifier needs to be converted to a base class.

ianmcburnie commented on docs/_includes/common/filter-menu-button.html:
```diff
@@ -0,0 +1,654 @@
+                                <span class="filter-menu-button__text">Item 1 with a long string that will wrap to another line</span>
+                            </div>
+                            <div class="filter-menu-button__item" role="menuitemcheckbox" aria-checked="false">
+                                <span class="filter-menu-button__status--inline">
```
@seangates @msendlakowski so this got merged with the invalid BEM still present? :-( A base class needs to be present, or this modifier needs to be converted to a base class.

seangates replied:
I'll get it fixed.
